firmware/TempoHandler.h

```diff
@@ -49,38 +50,33 @@ class TempoHandler
     inline void setHandleAlignEvent(void (*fptr)()) {
       tAlignCallback = fptr;
     }
-    void setPPQN(int ppqn) {
-      _ppqn = ppqn;
-    }
     void update() {
-      // Determine which source is selected for tempo
-      uint8_t new_source = _source;
-
-      if(Sync::detect()) {
-        new_source = TEMPO_SOURCE_SYNC;
-      } else if (_midi_clock_received_flag) { 
-        new_source = TEMPO_SOURCE_MIDI;
-      } else {
-        new_source = TEMPO_SOURCE_INTERNAL;
-      }
-
-      if (new_source != _source) {
-        _source = new_source;
-        _midi_clock_received_flag = 0;
-        tempo.reset();
+      static const unsigned second = 1000 * 1000;
+      static const unsigned timeout = 2 * second;
+
+      // If no clock has been output for some time, reset to internal source.
+      // Otherwise check sync.
+      if (micros() - _previous_clock_time > timeout) {
+        reset_clock_source();
+      }else if(Sync::detect()) {
+        if (_source != TEMPO_SOURCE_SYNC) {
+          reset_clock();
+        }
+        _source = TEMPO_SOURCE_SYNC;
+      }else if(_source == TEMPO_SOURCE_SYNC){
+        reset_clock_source();
       }
 
-
       switch(_source) {
         case TEMPO_SOURCE_INTERNAL:
           tempo.update_internal(*this, synth_params.speed);
           break;
-        case TEMPO_SOURCE_MIDI:
-          update_midi(midi_clock);
-          break;
         case TEMPO_SOURCE_SYNC:
           update_sync();
           break;
+        case TEMPO_SOURCE_MIDI:
+          // Nothing to do.
+          break;
```

Comment:
Since we now trigger directly on the midi callback and do the fallback to internal source based on time since last clock output, we don't need to do anything more when running from MIDI clock.

Reply:
we shouldn't fall back to the internal clock after a timeout. The only way to fall back should be to unplug and replug the sync in, or to power cycle the unit

Reply:
Removed